Serialize fewer archetypes

slavapestov opened the pull request:
The extended type of an extension, and the default type of an associated type, should both be stored as interface types. We prefer not to serialize archetypes because deserializing them is more expensive, since it requires constructing a GenericSignatureBuilder.

slavapestov committed:
AST: Extended types of extensions should be interface types

slavapestov commented:
@DougGregor this is a nice cleanup, but a few crashers regress in the GSB.

DougGregor commented:
FWIW, the first commit is pulled into https://github.com/apple/swift/pull/18859 on top of the `TypeResolution` work.

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov committed:
AST: Simpler implementation of DeclContext::getDeclaredInterfaceType() and getDeclaredTypeInContext()

slavapestov committed:
AST: Associated type defaults should be interface types

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test compiler performance

slavapestov commented on lib/AST/DeclContext.cpp:
```diff
@@ -101,59 +101,21 @@ GenericTypeParamType *DeclContext::getProtocolSelfType() const {
+    auto *NTD = ED->getExtendedNominal();
+    if (NTD == nullptr)
+      return ErrorType::get(ED->getASTContext());
+    return NTD->getDeclaredInterfaceType();
```
@DougGregor note that it would be nice if this returned the extended type instead to preserve sugar, but this function is called from inside validateExtension(), when we do name lookups while building the extension's generic signature...

DougGregor replied:
This slightly reduces dependencies; that's good.

DougGregor commented on lib/AST/DeclContext.cpp:
```diff
@@ -101,59 +101,21 @@ GenericTypeParamType *DeclContext::getProtocolSelfType() const {
 Type DeclContext::getDeclaredTypeInContext() const {
   if (auto *ED = dyn_cast<ExtensionDecl>(this))
-    return computeExtensionType(ED, DeclTypeKind::DeclaredTypeInContext);
+    return ED->mapTypeIntoContext(getDeclaredInterfaceType());
```
Much cleaner!

DougGregor commented on lib/Sema/TypeCheckType.cpp:
```diff
@@ -660,16 +660,6 @@ Type TypeChecker::applyGenericArguments(Type type,
-  // Cannot extend a bound generic type.
-  if (options.is(TypeResolverContext::ExtensionBinding)) {
-    if (!options.contains(TypeResolutionFlags::SilenceErrors)) {
-      diags.diagnose(loc, diag::extension_specialization,
```
Why remove this error?

slavapestov replied:
It's duplicated in validateExtension()

slavapestov committed:
AST: More robust differenceAndStoreConditionalRequirements()

Make sure we handle invalid extensions here, since calling
resolveExtension() won't always give us a generic signature.

This fixes a regression exposed by the next patch.

slavapestov commented:
@swift-ci Please test

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please test source compatibility

slavapestov commented:
@swift-ci Please smoke test macOS
